src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,655 @@
+  }
+
+  if (loglevelActiveFor(LOG_LEVEL_INFO))
+  {
+    String log;
+#ifdef P145_DEBUG
+    log = F("MQ-xx: Calibration with Rcal =  ");
+    log += currentRcal; // Calculated Rzero if calibration concentration is applied
+    log += F(" Rlast = ");
+    log += lastRcal;    // Rcal as calculated previous sample
+    addLog(LOG_LEVEL_INFO, log);
+#endif
+    if (doit)
+    {
+      log = F("MQ-xx: ***Calibrating*** Rzero =  ");
+      log += rcal;
+      addLog(LOG_LEVEL_INFO, log);
+    }
+  }
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the resistance of the sensor, ie. the property to be measured
+@param[in] val     Analog input value read from ADC [counts]
+@return The sensor resistance in Ohm
+@note   Uses hard coded constants:
+        MAX_ADC_VALUE Max range for ADC
+        VMAX          Analog  input voltage corresponding to MAX_ADC_VALUE
+        VCC           Voltage applied to the sensor-Rload combination
+*/
+/*****************************************************************************/
+float P145_data_struct::getResistance(float val)
+{
+  return ((MAX_ADC_VALUE * P145_VCC) / (P145_VMAX * val) - 1.0) * rload;
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the resistance RZero of the sensor for calibration purposes
+@param[in] rSensor    Sensor resistance value Rs
+@return The sensor resistance RZero in kOhm
+*/
+/*****************************************************************************/
+float P145_data_struct::getRZero(float rSensor)
+{
+  float newValue = rSensor;  // Default to current measured Rsensor
+  
+  if (newValue < 0.0) newValue = 0.0;  // Clip negative values
+  switch (algorithm)
+  {
+    case p145AlgA:  // MQ-135
+      newValue = (rSensor * pow((refLevel / sensordef.para), (1.0 / sensordef.parb)));
+      break;
+    case p145AlgB:  // Miquel5612 Exponential
+      if (sensordef.cleanRatio > 0.0)
+      {
+        newValue = (rSensor / sensordef.cleanRatio);
+      }
+      break;
+    case p145AlgC:  // Miquel5612 Linear
+      if (sensordef.cleanRatio > 0.0)
+      {
+        newValue = (rSensor / sensordef.cleanRatio);
+      }
+      break;
+    default:
+      if (sensordef.cleanRatio > 0.0)
+      {
+        newValue = (rSensor / sensordef.cleanRatio);
+      }
+      break;    
+  };
+  return (newValue);
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the corrected resistance RZero of the sensor for calibration
+        purposes
+@param[in] rSensor      Sensor resistance value Rs
+@param[in] temperature  The ambient air temperature
+@param[in] humidity     The relative humidity
+
+@return The corrected sensor resistance RZero in kOhm
+*/
+/*****************************************************************************/
+float P145_data_struct::getCorrectedRZero(float rSensor, float temperature, float humidity)
```

Comment:
Also a `const` function.

Reply:
Compiler does't like it to be const

Reply:
In that function you call `getRZero`.
As long as that function isn't `const`, all functions calling it cannot be `const`.
But as far as I can see, both functions do not change members.
So unless it is marked as a `static` function, then both should be able to set as `const`.